pkg/resourcemanager/controller/health/progressing/reconciler.go

```diff
@@ -173,19 +178,39 @@ func (r *Reconciler) reconcile(ctx context.Context, log logr.Logger, mr *resourc
 
 // checkProgressing checks whether the given object is progressing. It returns a bool indicating whether the object is
 // progressing, a reason for it if so and an error if the check failed.
-func checkProgressing(obj client.Object) (bool, string) {
+func (r *Reconciler) checkProgressing(ctx context.Context, obj client.Object) (bool, string, error) {
 	if obj.GetAnnotations()[resourcesv1alpha1.SkipHealthCheck] == "true" {
-		return false, ""
+		return false, "", nil
 	}
 
+	var (
+		progressing bool
+		reason      string
+	)
+
 	switch o := obj.(type) {
 	case *appsv1.Deployment:
-		return health.IsDeploymentProgressing(o)
+		progressing, reason = health.IsDeploymentProgressing(o)
+		if progressing {
+			return true, reason, nil
+		}
+
+		// health.IsDeploymentProgressing might return false even if there are still (terminating) pods in the system
+		// belonging to an older ReplicaSet of the Deployment, hence, we have to check for this explicitly.
+		exactNumberOfPods, err := health.DeploymentHasExactNumberOfPods(ctx, r.TargetClient, o)
+		if err != nil {
+			return progressing, reason, err
+		}
+		if !exactNumberOfPods {
+			return true, "there are still non-terminated old pods", nil
```

Comment:
Any reason to add here and not in `IsDeploymentProgressing`?

Reply:
`IsDeploymentProgressing` only operates on the `Deployment` object w/o the need of `context.Context` or `client.Client`, and I didn't want to change this. As discussed in https://github.com/gardener/gardener/issues/5850#issuecomment-1108251803, the majority of the cases are rather not depending on old pods to be terminating, so I decided for a dedicated function that can be used only for those "special scenarios".

Reply:
Ok, Thanks 👍